src/pages/index.js

```diff
@@ -33,60 +47,165 @@ const cardSection = new Section({
   }
 }, ".cards");
 const popupWithImage = new PopupWithImage('.modal_type_image');
-const api = new Api({
-  baseUrl: "https://around-api.en.tripleten-services.com/v1",
-  headers: {
-    authorization: "32400e4e-bdb6-4554-9f2b-339b0c55f4e6",
-    "Content-Type": "application/json"
-  }
-});
 
 
-// define functions
-function renderProfileDetails(){
-  const { name, description } = userInfo.getUserInfo();
-  nameInput.value = name;
-  jobInput.value = description;
+/* ------------------------------------------------------ */
+/*                    define functions                    */
+/* ------------------------------------------------------ */
+
+function renderProfileDetails(modal){
+  const { name, about } = userInfo.getUserInfo();
+  modal.setInputValues({name, about})
 };
 
 function handleCardAdd(data){
-  const { title, url } = data
-  const cardElement = createCard({title, url})
-  cardSection.addItem(cardElement)
-  cardModal.close()
+  this.renderLoading(true)
+  const { name, link } = data
+  api.addCard(name, link).then((data)=>{
+    cardSection.addItem(createCard(data));
+    formValidators['cardForm'].disableButton();
+    formValidators['cardForm'].resetInputs();
+    cardModal.close()
+  }).catch(err => {
+    console.error(err)
+  }).finally(()=>{
+    this.renderLoading(false)
+  });
 };
 
 function handleImageClick(image){
   const data = {
-    name: image.title,
-    link: image.url
+    name: image.name,
+    link: image.link
   }
   popupWithImage.open(data)
 };
 
+function handleDeleteClick(card){
+  deleteConfirmModal.open()
+  
+  deleteConfirmModal.setSubmitHandler(()=>{
+    deleteConfirmModal.renderLoading(true)
+    api.deleteCard(card.getCardId()).then(()=>{
+      card.delete()
+      deleteConfirmModal.close()
+    }).catch(err => {
+      console.error(err)
+    }).finally(()=>{
+      deleteConfirmModal.renderLoading(false)
+    })
+  })
+}
+
 function createCard(cardData){
-  const card = new Card(cardData, '.card-template', handleImageClick);
+  const card = new Card(cardData, '.card-template', handleImageClick, handleDeleteClick, handleLikeClick, handleDislikeClick);
   return card.generateCard()
 };
 
 function handleProfileFormSubmit(data){
-  userInfo.setUserInfo(data);
+  this.renderLoading(true)
+  const {name, about} = data;
+  api.updateProfile(name, about).then(data => {
+    return data
+  }).then((info) => {
+    userInfo.setUserInfo(info)
+    formValidators['profileForm'].disableButton();
+    formValidators['profileForm'].resetInputs();
+    profileModal.close()
+  }).catch(err => {
+    console.error(err)
+  }).finally(()=>{
+    this.renderLoading(false)
```

Comment:
**Could be improved**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```